File: server/core/job_processor.py
"""
Background job processor.

Architecture
------------
- A single JobProcessor instance is created at app startup.
- A polling loop runs every POLL_INTERVAL seconds and picks up PENDING jobs.
- Each job is processed with a semaphore (MAX_CONCURRENT_DOWNLOADS).
- Every job requires a pre-selected stream from the UI (no auto-pick).
- The pipeline per job:
    1. Deserialize pre-selected stream + media info
    2. Resolve all RD download URLs (season packs get every episode file)
    3. Download to staging area (streamed, chunked, with progress)
    4. Organise into library (Plex-compatible naming)
    5. Download TMDB poster into the central posters directory
    6. Mark COMPLETE
"""
from __future__ import annotations

import asyncio
import logging
import re
from dataclasses import replace as dc_replace
from pathlib import Path
from typing import Optional

import aiofiles
import httpx

from ..config import settings
from ..database import (
    JobStatus,
    append_log,
    get_job,
    get_pending_jobs,
    update_job,
)
from .media_organizer import MediaOrganizer
from ..clients.torrentio_client import StreamResult
from ..clients.realdebrid_client import RealDebridError
from ..clients.tmdb_client import MediaInfo

logger = logging.getLogger(__name__)

POLL_INTERVAL = 5   # seconds between queue polls

# Video file extensions used when filtering RD links for season packs
_VIDEO_EXTS = {".mkv", ".mp4", ".avi", ".m2ts", ".ts", ".mov", ".wmv", ".m4v"}

# Maximum download time per job (2 hours)
_MAX_DOWNLOAD_SECONDS = 2 * 60 * 60


class JobProcessor:
    def __init__(self):
        from .. import state

        self._sem = asyncio.Semaphore(settings.MAX_CONCURRENT_DOWNLOADS)
        self._active: dict[str, asyncio.Task] = {}
        self._running = False

        # Reuse the app-level singletons populated by main.lifespan()
        self._tmdb = state.tmdb
        self._torrentio = state.torrentio
        self._rd = state.rd
        self._organizer = MediaOrganizer()

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self):
        self._running = True
        asyncio.create_task(self._poll_loop())
        logger.info("JobProcessor started")

    def stop(self):
        self._running = False
        logger.info("JobProcessor stopping -- %d active jobs", len(self._active))

    def cancel_job(self, job_id: str) -> bool:
        task = self._active.get(job_id)
        if task:
            task.cancel()
            return True
        return False

    # ------------------------------------------------------------------
    # Poll loop
    # ------------------------------------------------------------------

    async def _poll_loop(self):
        while self._running:
            try:
                pending = await get_pending_jobs()
                for job in pending:
                    jid = job["id"]
                    if jid not in self._active:
                        task = asyncio.create_task(self._process(jid))
                        self._active[jid] = task
                        task.add_done_callback(lambda t, j=jid: self._active.pop(j, None))
            except Exception:
                logger.exception("Error in job poll loop")
            await asyncio.sleep(POLL_INTERVAL)

    # ------------------------------------------------------------------
    # Pipeline entry
    # ------------------------------------------------------------------

    async def _process(self, job_id: str):
        async with self._sem:
            try:
                await asyncio.wait_for(
                    self._run_pipeline(job_id),
                    timeout=_MAX_DOWNLOAD_SECONDS,
                )
            except asyncio.CancelledError:
                await update_job(job_id, status=JobStatus.CANCELLED)
                await _log(job_id, "Job cancelled")
                logger.info("Job %s cancelled", job_id)
                await self._cleanup_staging(job_id)
            except asyncio.TimeoutError:
                await update_job(job_id, status=JobStatus.FAILED, error="Download timed out")
                await _log(job_id, "ERROR: Download timed out")
                await self._cleanup_staging(job_id)
            except Exception as exc:
                logger.exception("Job %s failed: %s", job_id, exc)
                await update_job(job_id, status=JobStatus.FAILED, error=str(exc))
                await _log(job_id, f"ERROR: {exc}")
                await self._cleanup_staging(job_id)

    async def _cleanup_staging(self, job_id: str):
        """Remove any partially downloaded files in the staging directory."""
        staging_dir = Path(settings.DOWNLOADS_DIR)
        if not staging_dir.exists():
            return
        for f in staging_dir.iterdir():
            if f.is_file() and job_id[:8] in f.name:
                try:
                    f.unlink()
                    logger.info("Cleaned up staging file: %s", f)
                except Exception:
                    pass

    async def _run_pipeline(self, job_id: str):
        import json as _json
        job = await get_job(job_id)
        if not job:
            return

        # All jobs must have pre-selected stream data
        if not job.get("stream_data"):
            raise ValueError("No stream selected — job requires a pre-selected stream from the UI")

        sd = _json.loads(job["stream_data"])
        media_d = sd["media"]
        stream_d = sd["stream"]

        media = MediaInfo(
            title=media_d["title"],
            year=media_d.get("year"),
            imdb_id=media_d.get("imdb_id"),
            tmdb_id=media_d.get("tmdb_id"),
            type=media_d.get("type", "movie"),
            season=media_d.get("season"),
            episode=media_d.get("episode"),
            is_anime=media_d.get("is_anime", False),
            episode_titles=media_d.get("episode_titles", {}),
            poster_path=media_d.get("poster_path"),
        )

        best = StreamResult(
            name=stream_d["name"],
            info_hash=stream_d.get("info_hash"),
            download_url=stream_d.get("download_url"),
            size_bytes=stream_d.get("size_bytes"),
            seeders=stream_d.get("seeders", 0),
            is_cached_rd=stream_d.get("is_cached_rd", False),
            magnet=stream_d.get("magnet"),
            file_idx=stream_d.get("file_idx"),
        )

        await update_job(
            job_id,
            title=media.title, year=media.year, imdb_id=media.imdb_id,
            type=media.type, season=media.season, episode=media.episode,
            torrent_name=best.name[:120],
            status=JobStatus.FOUND,
        )
        await _log(job_id, f"Starting: {media.display_name}")
        await _log(job_id, f"Stream: {best.name[:80]}")

        # Resolve download files via RD
        download_files = await self._resolve_rd_files(job_id, best, media)

        # Download, organise, save poster
        staging_dir = Path(settings.DOWNLOADS_DIR)
        staging_dir.mkdir(parents=True, exist_ok=True)
        await self._download_and_organize(job_id, download_files, media, staging_dir)

    # ------------------------------------------------------------------
    # Resolve RD download URLs
    # ------------------------------------------------------------------

    async def _resolve_rd_files(
        self, job_id: str, best: StreamResult, media: MediaInfo
    ) -> list[tuple[str, Optional[int]]]:
        """
        Return list of (cdn_url, size_bytes) to download.
        - Single episode/movie: one entry (uses Torrentio direct URL if cached)
        - Season pack (TV/anime, no specific episode): ALL video files in torrent
        """
        is_season_pack = media.type in ("tv", "anime") and media.episode is None

        if best.is_cached_rd and best.download_url and not is_season_pack:
            await _log(job_id, "Using pre-resolved RD URL (instant cache)")
            return [(best.download_url, best.size_bytes)]

        # Go through RD API to get all file links
        magnet = best.magnet
        if not magnet and best.info_hash:
            magnet = f"magnet:?xt=urn:btih:{best.info_hash}"
        if not magnet:
            raise ValueError("No magnet link available for this stream")

        await update_job(job_id, status=JobStatus.ADDING_TO_RD)
        await _log(job_id, "Adding to Real-Debrid ...")
        torrent_id = await self._rd.add_magnet(magnet)
        await self._rd.select_all_files(torrent_id)
        await update_job(job_id, rd_torrent_id=torrent_id, status=JobStatus.WAITING_FOR_RD)
        await _log(job_id, f"Waiting for RD (id={torrent_id}) ...")

        async def _rd_progress(pct: int):
            await _log(job_id, f"RD progress: {pct}%")
            await update_job(job_id, progress=pct / 100 * 0.3)

        rd_links = await self._rd.wait_until_downloaded(torrent_id, on_progress=_rd_progress)
        await _log(job_id, f"RD ready -- unrestricting {len(rd_links)} link(s) ...")
        unrestricted = await self._rd.unrestrict_all(rd_links)

        if is_season_pack:
            # Filter to video files; fall back to all if filter is too aggressive
            video_files = [
                (u, s) for u, s in unrestricted
                if _is_video_url(u) or (s or 0) > 50 * 1024 * 1024
            ]
            files = video_files or unrestricted
            # Sort by filename so episodes come out in order
            files.sort(key=lambda x: x[0].split("?")[0].split("/")[-1].lower())
            await _log(job_id, f"Season pack: {len(files)} episode file(s)")
            return files
        else:
            unrestricted.sort(key=lambda x: x[1] or 0, reverse=True)
            return [unrestricted[0]]

    # ------------------------------------------------------------------
    # Download + organise
    # ------------------------------------------------------------------

    async def _download_and_organize(
        self,
        job_id: str,
        files: list[tuple[str, Optional[int]]],
        media: MediaInfo,
        staging_dir: Path,
    ) -> None:
        total_size = sum(s or 0 for _, s in files) or None
        await update_job(job_id, status=JobStatus.DOWNLOADING, size_bytes=total_size)

        last_final_path: Optional[Path] = None

        for i, (dl_url, fl_size) in enumerate(files):
            file_name = _filename_from_url(dl_url) or f"{job_id}_{i}.mkv"
            staging_path = staging_dir / file_name

            prefix = f"[{i+1}/{len(files)}] " if len(files) > 1 else ""
            await _log(job_id, f"{prefix}Downloading: {file_name}")
            await self._download(job_id, dl_url, staging_path, fl_size)

            # For season packs, detect episode number from filename
            if media.episode is None and media.type in ("tv", "anime"):
                ep_num = _episode_from_filename(file_name)
                ep_media = dc_replace(media, episode=ep_num) if ep_num is not None else media
            else:
                ep_media = media

            await update_job(job_id, status=JobStatus.ORGANIZING)
            final_path = self._organizer.organize(staging_path, ep_media)
            await _log(job_id, f"Organised -> {final_path}")
            last_final_path = final_path

        if last_final_path:
            await update_job(
                job_id,
                status=JobStatus.COMPLETE,
                file_path=str(last_final_path),
                progress=1.0,
                downloaded_bytes=total_size or 0,
            )
            await _log(job_id, f"Done ({len(files)} file(s))")
            await _save_poster(media, last_final_path)

    # ------------------------------------------------------------------
    # Downloader
    # ------------------------------------------------------------------

    async def _download(
        self,
        job_id: str,
        url: str,
        dest: Path,
        total_size: Optional[int],
        _max_connect_retries: int = 3,
    ) -> None:
        downloaded = 0
        chunk_size = settings.CHUNK_SIZE

        # Retry the initial connection (CDN URLs can fail on first contact)
        last_exc: Optional[Exception] = None
        for attempt in range(_max_connect_retries):
            try:
                async with httpx.AsyncClient(follow_redirects=True, timeout=None) as client:
                    async with client.stream("GET", url) as response:
                        response.raise_for_status()

                        cl = response.headers.get("content-length")
                        if cl and cl.isdigit():
                            total_size = int(cl)
                            await update_job(job_id, size_bytes=total_size)

                        async with aiofiles.open(dest, "wb") as fh:
                            async for chunk in response.aiter_bytes(chunk_size):
                                await fh.write(chunk)
                                downloaded += len(chunk)

                                if total_size and total_size > 0:
                                    pct = 0.3 + (downloaded / total_size) * 0.7
                                else:
                                    pct = 0.5

                                await update_job(
                                    job_id,
                                    downloaded_bytes=downloaded,
                                    progress=min(pct, 0.99),
                                )

                await _log(job_id, f"Download complete: {downloaded / 1024**3:.2f} GB")
                return  # success
            except (httpx.ConnectError, httpx.ConnectTimeout) as exc:
                last_exc = exc
                delay = 2 * (2 ** attempt)
                await _log(
                    job_id,
                    f"Download connect failed (attempt {attempt + 1}/{_max_connect_retries}) "
                    f"— retrying in {delay}s",
                )
                await asyncio.sleep(delay)

        raise last_exc  # type: ignore[misc]


# ---------------------------------------------------------------------------
# Module-level helpers
# ---------------------------------------------------------------------------

def _filename_from_url(url: str) -> Optional[str]:
    try:
        import urllib.parse
        path = url.split("?")[0].rstrip("/")
        name = urllib.parse.unquote(path.split("/")[-1])
        name = re.sub(r"[?&=].*", "", name)
        if name and "." in name:
            return name
    except Exception:
        pass
    return None


def _is_video_url(url: str) -> bool:
    name = url.split("?")[0].rstrip("/").split("/")[-1].lower()
    return any(name.endswith(ext) for ext in _VIDEO_EXTS)


def _episode_from_filename(name: str) -> Optional[int]:
    """Parse episode number from filenames.

    Supports multiple patterns:
      - S01E03
      - E03 or Ep03
      - Episode 03
      - - 03 - (number between dashes, common in anime)
    """
    # Standard S##E## pattern
    m = re.search(r"[Ss]\d{1,2}[Ee](\d{1,3})", name)
    if m:
        return int(m.group(1))
    # E## or Ep## pattern
    m = re.search(r"(?:^|[\s._-])[Ee]p?(\d{1,3})(?:[\s._\-\[]|$)", name)
    if m:
        return int(m.group(1))
    # " - 03 - " pattern (common in anime)
    m = re.search(r"\s-\s(\d{1,3})\s-\s", name)
    if m:
        return int(m.group(1))
    # " - 03." or " - 03 [" pattern (anime at end)
    m = re.search(r"\s-\s(\d{1,3})(?:\.\w{3}$|\s*\[)", name)
    if m:
        return int(m.group(1))
    return None


async def _save_poster(media: MediaInfo, final_path: Path) -> None:
    """Download the TMDB poster to the central data/posters directory."""
    if not media.poster_url:
        return

    posters_dir = Path(settings.POSTERS_DIR)
    posters_dir.mkdir(parents=True, exist_ok=True)

    # Build the poster key
    if media.type == "movie":
        key = f"{media.title} ({media.year})" if media.year else media.title
    else:
        key = media.title

    poster_file = posters_dir / f"{_safe_poster_key(key)}.jpg"

    if poster_file.exists():
        return

    try:
        async with httpx.AsyncClient(timeout=20, follow_redirects=True) as client:
            r = await client.get(media.poster_url)
            r.raise_for_status()
            async with aiofiles.open(poster_file, "wb") as f:
                await f.write(r.content)
        logger.info("Saved poster -> %s", poster_file)
    except Exception as exc:
        logger.warning("Could not download poster: %s", exc)


def _safe_poster_key(s: str) -> str:
    """Strip characters that are illegal in Windows filenames."""
    return re.sub(r'[\\/:*?"<>|]', "_", s).strip()


async def _log(job_id: str, msg: str) -> None:
    logger.info("[job %s] %s", job_id[:8], msg)
    await append_log(job_id, msg)

File: app/MediaDownloader/MediaDownloader.Server/Core/JobProcessor.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using MediaDownloader.Server.Clients;
using MediaDownloader.Server.Configuration;
using MediaDownloader.Server.Data;

namespace MediaDownloader.Server.Core;

/// <summary>
/// Background job processor — polls for pending jobs and runs the download pipeline.
/// Mirrors server/core/job_processor.py.
/// </summary>
public class JobProcessor : BackgroundService
{
    private static readonly HashSet<string> VideoExts = [".mkv", ".mp4", ".avi", ".m2ts", ".ts", ".mov", ".wmv", ".m4v"];
    private const int PollInterval = 5;
    private const int MaxDownloadSeconds = 2 * 60 * 60;

    private readonly DbService _db;
    private readonly ServerSettings _settings;
    private readonly ILogger<JobProcessor> _logger;
    private readonly SemaphoreSlim _sem;
    private readonly HashSet<string> _active = [];
    private readonly object _activeLock = new();

    // Mutable references — updated when settings change
    public TmdbClient? Tmdb { get; set; }
    public TorrentioClient? Torrentio { get; set; }
    public RealDebridClient? Rd { get; set; }

    public JobProcessor(DbService db, ServerSettings settings, ILogger<JobProcessor> logger)
    {
        _db = db;
        _settings = settings;
        _logger = logger;
        _sem = new SemaphoreSlim(settings.MaxConcurrentDownloads);
    }

    public bool CancelJob(string jobId)
    {
        // Jobs run as fire-and-forget tasks; cancellation is done via DB status check in the pipeline
        return false;
    }

    protected override async Task ExecuteAsync(CancellationToken ct)
    {
        _logger.LogInformation("JobProcessor started");
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var pending = await _db.GetPendingJobsAsync();
                foreach (var job in pending)
                {
                    var jid = job["id"]?.ToString() ?? "";
                    bool shouldStart;
                    lock (_activeLock) { shouldStart = _active.Add(jid); }
                    if (shouldStart)
                    {
                        _ = Task.Run(async () =>
                        {
                            try { await ProcessJobAsync(jid, ct); }
                            finally { lock (_activeLock) { _active.Remove(jid); } }
                        }, ct);
                    }
                }
            }
            catch (Exception ex) { _logger.LogError(ex, "Error in job poll loop"); }
            await Task.Delay(TimeSpan.FromSeconds(PollInterval), ct);
        }
    }

    private async Task ProcessJobAsync(string jobId, CancellationToken ct)
    {
        await _sem.WaitAsync(ct);
        try
        {
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            cts.CancelAfter(TimeSpan.FromSeconds(MaxDownloadSeconds));
            await RunPipelineAsync(jobId, cts.Token);
        }
        catch (OperationCanceledException)
        {
            await _db.UpdateJobAsync(jobId, new() { ["status"] = JobStatus.Cancelled });
            await LogAsync(jobId, "Job cancelled");
            await CleanupStagingAsync(jobId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Job {JobId} failed", jobId[..8]);
            await _db.UpdateJobAsync(jobId, new() { ["status"] = JobStatus.Failed, ["error"] = ex.Message });
            await LogAsync(jobId, $"ERROR: {ex.Message}");
            await CleanupStagingAsync(jobId);
        }
        finally { _sem.Release(); }
    }

    private async Task CleanupStagingAsync(string jobId)
    {
        if (!Directory.Exists(_settings.DownloadsDir)) return;
        foreach (var f in Directory.EnumerateFiles(_settings.DownloadsDir))
        {
            if (Path.GetFileName(f).Contains(jobId[..8]))
            {
                try { File.Delete(f); } catch { }
            }
        }
    }

    private async Task RunPipelineAsync(string jobId, CancellationToken ct)
    {
        var job = await _db.GetJobAsync(jobId);
        if (job == null) return;

        var streamDataStr = job["stream_data"]?.ToString();
        if (string.IsNullOrEmpty(streamDataStr))
            throw new InvalidOperationException("No stream selected — job requires a pre-selected stream from the UI");

        var sd = JsonSerializer.Deserialize<JsonElement>(streamDataStr);
        var mediaD = sd.GetProperty("media");
        var streamD = sd.GetProperty("stream");

        var media = new MediaInfo
        {
            Title = GetStr(mediaD, "title") ?? "",
            Year = GetInt(mediaD, "year"),
            ImdbId = GetStr(mediaD, "imdb_id"),
            TmdbId = GetInt(mediaD, "tmdb_id"),
            Type = GetStr(mediaD, "type") ?? "movie",
            Season = GetInt(mediaD, "season"),
            Episode = GetInt(mediaD, "episode"),
            IsAnime = mediaD.TryGetProperty("is_anime", out var ia) && ia.GetBoolean(),
            PosterPath = GetStr(mediaD, "poster_path"),
        };

        // Parse episode titles
        if (mediaD.TryGetProperty("episode_titles", out var et) && et.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in et.EnumerateObject())
            {
                if (int.TryParse(prop.Name, out var epNum))
                    media.EpisodeTitles[epNum] = prop.Value.GetString() ?? "";
            }
        }

        var best = new StreamResult
        {
            Name = GetStr(streamD, "name") ?? "",
            InfoHash = GetStr(streamD, "info_hash"),
            DownloadUrl = GetStr(streamD, "download_url"),
            SizeBytes = GetLong(streamD, "size_bytes"),
            Seeders = GetInt(streamD, "seeders") ?? 0,
            IsCachedRd = streamD.TryGetProperty("is_cached_rd", out var cr) && cr.GetBoolean(),
            Magnet = GetStr(streamD, "magnet"),
            FileIdx = GetInt(streamD, "file_idx"),
        };

        await _db.UpdateJobAsync(jobId, new()
        {
            ["title"] = media.Title, ["year"] = (object?)media.Year ?? DBNull.Value,
            ["imdb_id"] = (object?)media.ImdbId ?? DBNull.Value, ["type"] = media.Type,
            ["season"] = (object?)media.Season ?? DBNull.Value, ["episode"] = (object?)media.Episode ?? DBNull.Value,
            ["torrent_name"] = best.Name.Length > 120 ? best.Name[..120] : best.Name,
            ["status"] = JobStatus.Found,
        });
        await LogAsync(jobId, $"Starting: {media.DisplayName}");
        await LogAsync(jobId, $"Stream: {(best.Name.Length > 80 ? best.Name[..80] : best.Name)}");

        var downloadFiles = await ResolveRdFilesAsync(jobId, best, media, ct);

        Directory.CreateDirectory(_settings.DownloadsDir);
        await DownloadAndOrganizeAsync(jobId, downloadFiles, media, ct);
    }

    private async Task<List<(string url, long? size)>> ResolveRdFilesAsync(
        string jobId, StreamResult best, MediaInfo media, CancellationToken ct)
    {
        var isSeasonPack = media.Type is "tv" or "anime" && !media.Episode.HasValue;

        if (best.IsCachedRd && best.DownloadUrl != null && !isSeasonPack)
        {
            await LogAsync(jobId, "Using pre-resolved RD URL (instant cache)");
            return [(best.DownloadUrl, best.SizeBytes)];
        }

        var magnet = best.Magnet ?? (best.InfoHash != null ? $"magnet:?xt=urn:btih:{best.InfoHash}" : null);
        if (magnet == null) throw new InvalidOperationException("No magnet link available for this stream");

        await _db.UpdateJobAsync(jobId, new() { ["status"] = JobStatus.AddingToRd });
        await LogAsync(jobId, "Adding to Real-Debrid ...");
        var torrentId = await Rd!.AddMagnetAsync(magnet, ct);
        await Rd.SelectAllFilesAsync(torrentId, ct);
        await _db.UpdateJobAsync(jobId, new() { ["rd_torrent_id"] = torrentId, ["status"] = JobStatus.WaitingForRd });
        await LogAsync(jobId, $"Waiting for RD (id={torrentId}) ...");

        var rdLinks = await Rd.WaitUntilDownloadedAsync(torrentId, async pct =>
        {
            await LogAsync(jobId, $"RD progress: {pct}%");
            await _db.UpdateJobAsync(jobId, new() { ["progress"] = pct / 100.0 * 0.3 });
        }, ct: ct);

        await LogAsync(jobId, $"RD ready -- unrestricting {rdLinks.Count} link(s) ...");
        var unrestricted = await Rd.UnrestrictAllAsync(rdLinks, ct);

        if (isSeasonPack)
        {
            var videoFiles = unrestricted
                .Where(x => IsVideoUrl(x.url) || (x.size ?? 0) > 50 * 1024 * 1024)
                .ToList();
            var files = videoFiles.Count > 0 ? videoFiles : unrestricted;
            files.Sort((a, b) => string.Compare(
                a.url.Split('?')[0].Split('/')[^1],
                b.url.Split('?')[0].Split('/')[^1],
                StringComparison.OrdinalIgnoreCase));
            await LogAsync(jobId, $"Season pack: {files.Count} episode file(s)");
            return files;
        }

        unrestricted.Sort((a, b) => (b.size ?? 0).CompareTo(a.size ?? 0));
        return [unrestricted[0]];
    }

    private async Task DownloadAndOrganizeAsync(
        string jobId, List<(string url, long? size)> files, MediaInfo media, CancellationToken ct)
    {
        var totalSize = files.Sum(f => f.size ?? 0);
        await _db.UpdateJobAsync(jobId, new()
        {
            ["status"] = JobStatus.Downloading,
            ["size_bytes"] = totalSize > 0 ? totalSize : (object)DBNull.Value,
        });

        string? lastFinalPath = null;
        var organizer = new MediaOrganizer(_settings, _logger);

        for (int i = 0; i < files.Count; i++)
        {
            var (dlUrl, flSize) = files[i];
            var fileName = FilenameFromUrl(dlUrl) ?? $"{jobId}_{i}.mkv";
            var stagingPath = Path.Combine(_settings.DownloadsDir, fileName);

            var prefix = files.Count > 1 ? $"[{i + 1}/{files.Count}] " : "";
            await LogAsync(jobId, $"{prefix}Downloading: {fileName}");
            await DownloadFileAsync(jobId, dlUrl, stagingPath, flSize, ct);

            // For season packs, detect episode number from filename
            var epMedia = media;
            if (!media.Episode.HasValue && media.Type is "tv" or "anime")
            {
                var epNum = EpisodeFromFilename(fileName);
                if (epNum.HasValue) epMedia = media.WithEpisode(epNum);
            }

            await _db.UpdateJobAsync(jobId, new() { ["status"] = JobStatus.Organizing });
            var finalPath = organizer.Organize(stagingPath, epMedia);
            await LogAsync(jobId, $"Organised -> {finalPath}");
            lastFinalPath = finalPath;
        }

        if (lastFinalPath != null)
        {
            await _db.UpdateJobAsync(jobId, new()
            {
                ["status"] = JobStatus.Complete,
                ["file_path"] = lastFinalPath,
                ["progress"] = 1.0,
                ["downloaded_bytes"] = totalSize > 0 ? totalSize : (object)DBNull.Value,
            });
            await LogAsync(jobId, $"Done ({files.Count} file(s))");
            await SavePosterAsync(media);
        }
    }

    private async Task DownloadFileAsync(
        string jobId, string url, string dest, long? totalSize, CancellationToken ct, int maxRetries = 3)
    {
        long downloaded = 0;
        Exception? lastExc = null;

        for (int attempt = 0; attempt < maxRetries; attempt++)
        {
            try
            {
                using var client = new HttpClient(new HttpClientHandler { AllowAutoRedirect = true })
                    { Timeout = Timeout.InfiniteTimeSpan };
                using var response = await client.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct);
                response.EnsureSuccessStatusCode();

                if (response.Content.Headers.ContentLength.HasValue)
                {
                    totalSize = response.Content.Headers.ContentLength.Value;
                    await _db.UpdateJobAsync(jobId, new() { ["size_bytes"] = totalSize });
                }

                await using var stream = await response.Content.ReadAsStreamAsync(ct);
                await using var fileStream = new FileStream(dest, FileMode.Create, FileAccess.Write, FileShare.None, _settings.ChunkSize);
                var buffer = new byte[_settings.ChunkSize];

                while (true)
                {
                    var read = await stream.ReadAsync(buffer, ct);
                    if (read == 0) break;
                    await fileStream.WriteAsync(buffer.AsMemory(0, read), ct);
                    downloaded += read;

                    var pct = totalSize > 0 ? 0.3 + (double)downloaded / totalSize.Value * 0.7 : 0.5;
                    await _db.UpdateJobAsync(jobId, new()
                    {
                        ["downloaded_bytes"] = downloaded,
                        ["progress"] = Math.Min(pct, 0.99),
                    });
                }

                await LogAsync(jobId, $"Download complete: {downloaded / (1024.0 * 1024 * 1024):F2} GB");
                return;
            }
            catch (HttpRequestException ex)
            {
                lastExc = ex;
                var delay = 2 * (int)Math.Pow(2, attempt);
                await LogAsync(jobId, $"Download connect failed (attempt {attempt + 1}/{maxRetries}) — retrying in {delay}s");
                await Task.Delay(TimeSpan.FromSeconds(delay), ct);
            }
        }

        throw lastExc!;
    }

    private async Task SavePosterAsync(MediaInfo media)
    {
        if (media.PosterUrl == null) return;

        Directory.CreateDirectory(_settings.PostersDir);
        var key = media.Type == "movie"
            ? (media.Year.HasValue ? $"{media.Title} ({media.Year})" : media.Title)
            : media.Title;
        var safeKey = Regex.Replace(key, @"[\\/:*?""<>|]", "_").Trim();
        var posterFile = Path.Combine(_settings.PostersDir, $"{safeKey}.jpg");
        if (File.Exists(posterFile)) return;

        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(20) };
            var data = await client.GetByteArrayAsync(media.PosterUrl);
            await File.WriteAllBytesAsync(posterFile, data);
            _logger.LogInformation("Saved poster -> {File}", posterFile);
        }
        catch (Exception ex) { _logger.LogWarning("Could not download poster: {Error}", ex.Message); }
    }

    private async Task LogAsync(string jobId, string msg)
    {
        _logger.LogInformation("[job {Id}] {Msg}", jobId[..8], msg);
        await _db.AppendLogAsync(jobId, msg);
    }

    // ------------------------------------------------------------------
    // Static helpers
    // ------------------------------------------------------------------

    private static string? FilenameFromUrl(string url)
    {
        try
        {
            var path = url.Split('?')[0].TrimEnd('/');
            var name = Uri.UnescapeDataString(path.Split('/')[^1]);
            name = Regex.Replace(name, @"[?&=].*", "");
            return !string.IsNullOrEmpty(name) && name.Contains('.') ? name : null;
        }
        catch { return null; }
    }

    private static bool IsVideoUrl(string url)
    {
        var name = url.Split('?')[0].TrimEnd('/').Split('/')[^1].ToLowerInvariant();
        return VideoExts.Any(name.EndsWith);
    }

    internal static int? EpisodeFromFilename(string name)
    {
        var m = Regex.Match(name, @"[Ss]\d{1,2}[Ee](\d{1,3})");
        if (m.Success) return int.Parse(m.Groups[1].Value);
        m = Regex.Match(name, @"(?:^|[\s._-])[Ee]p?(\d{1,3})(?:[\s._\-\[]|$)");
        if (m.Success) return int.Parse(m.Groups[1].Value);
        m = Regex.Match(name, @"\s-\s(\d{1,3})\s-\s");
        if (m.Success) return int.Parse(m.Groups[1].Value);
        m = Regex.Match(name, @"\s-\s(\d{1,3})(?:\.\w{3}$|\s*\[)");
        if (m.Success) return int.Parse(m.Groups[1].Value);
        return null;
    }

    private static string? GetStr(JsonElement el, string prop) =>
        el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
    private static int? GetInt(JsonElement el, string prop) =>
        el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetInt32() : null;
    private static long? GetLong(JsonElement el, string prop) =>
        el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetInt64() : null;
}
